### PR TITLE
from chris coderev

### PR DESCRIPTION
**Pull request by earonesty**
Address all comments from chris initial-commit coderev.

### CODE REVIEW

File: omen2/codegen.py [hunk 1]
```diff
@@ -41,29 +58,48 @@ def gen_class(out, name, dbtab: "DbTable"):
             out: file stream
             name: table name
             dbtab: table model
+
```
Comment (higleyc) [inline omen2/codegen.py:61]: thanks, easier to understand

File: omen2/object.py [hunk 2]
```diff
@@ -339,13 +350,22 @@ def __enter__(self):
...
             if not typ:
-                self._commit()
+                # see self._atomic_update() for why this works
+                saved = self.__dict__  # pylint: disable=access-member-before-definition
+                try:
+                    self._commit()
+                except Exception:
+                    self.__dict__ = saved
+                    raise
+        except OmenRollbackError:
```
Comment (higleyc) [inline omen2/object.py:367]: What I meant with this is that the `with` user would throw `OmenRollbackError`, which would come through in `typ`, and then we could return `True` here to suppress

Comment (earonesty) [reply omen2/object.py:370]: yeah, i knew what you meant, i just wasnt thinking.  will add a test

File: omen2/codegen.py [hunk 3]
```diff
@@ -41,29 +58,48 @@ def gen_class(out, name, dbtab: "DbTable"):
...
             print(name_and_type, file=out, end="")
+
             if col.default or not col.notnull:
+                # derive default value from the db default value
                 if col.default:
                     try:
                         defval = pytype(col.default)
-                    except ValueError:
+                        # check valid python
+                        eval(str(defval))  # pylint: disable=eval-used
```
[the remark this answers is not in the document]
Comment (earonesty) [reply omen2/codegen.py:100]: it's to check whether the statement can be evaluated successfully.   the canonical example is:  `default localtime()` ... which is syntactically fine... but isn't something python understands.

i could probably use AST to verify that it's a constant... ?   eval just seemed easier.

File: omen2/codegen.py [hunk 4]
```diff
@@ -4,18 +4,20 @@
...
-from typing import TYPE_CHECKING
 
-if TYPE_CHECKING:
-    from notanorm import DbTable
+from notanorm import DbTable, DbCol
+
+from omen2.types import default_type
 
 
 class CodeGen:
```
[the remark this answers is not in the document]
Comment (earonesty) [reply omen2/codegen.py:13]: seems useful.

File: omen2/codegen.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (earonesty) [reply omen2/codegen.py:100]: added to "from coderev"

File: omen2/object.py (hunk 2 above)
Comment (earonesty) [reply omen2/object.py]: updated

File: omen2/codegen.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (earonesty) [reply omen2/codegen.py:13]: added example to the README

**Pull request merged**